doc/man/man8/slapmodify.8
.TH SLAPMODIFY 8C "RELEASEDATE" "OpenLDAP LDVERSION"
.\" Copyright 1998-2020 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.\" $OpenLDAP$
.SH NAME
slapmodify \- Modify entries in a SLAPD database
.SH SYNOPSIS
.B SBINDIR/slapmodify
[\c
.BI \-b \ suffix\fR]
[\c
.BR \-c ]
[\c
.BI \-d \ debug-level\fR]
[\c
.BI \-f \ slapd.conf\fR]
[\c
.BI \-F \ confdir\fR]
[\c
.BR \-g ]
[\c
.BI \-j \ lineno\fR]
[\c
.BI \-l \ ldif-file\fR]
[\c
.BI \-n \ dbnum\fR]
[\c
.BI \-o \ option\fR[ = value\fR]]
[\c
.BR \-q ]
[\c
.BR \-s ]
[\c
.BI \-S \ SID\fR]
[\c
.BR \-u ]
[\c
.BR \-v ]
[\c
.BR \-w ]
.SH DESCRIPTION
.LP
.B Slapmodify
is used to apply modifications specified in LDAP Directory Interchange Format
(LDIF) to a
.BR slapd (8)
database.
It opens the given database determined by the database number or
suffix and performs modifications corresponding to the provided LDIF to
the database.
Databases configured as
.B subordinate
of this one are also updated, unless \fB\-g\fP is specified.
The LDIF input is read from standard input or the specified file.

All files eventually created by
.BR slapmodify
will belong to the identity
.BR slapmodify
is run as, so make sure you either run
.BR slapmodify
with the same identity
.BR slapd (8)
will be run as (see option
.B \-u
in
.BR slapd (8)),
or change file ownership before running
.BR slapd (8).

Note: slapmodify will also perform the relevant indexing whilst modifying the database if
any are configured. For specific details, please see
.BR slapindex (8).
.SH OPTIONS
.TP
.BI \-b \ suffix
Use the specified \fIsuffix\fR to determine which database to
add entries to.  The \fB\-b\fP cannot be used in conjunction
with the
.B \-n
option.
.TP
.B \-c
enable continue (ignore errors) mode.
.TP
.BI \-d \ debug-level
enable debugging messages as defined by the specified
.IR debug-level ;
see
.BR slapd (8)
for details.
.TP
.BI \-f \ slapd.conf
specify an alternative
.BR slapd.conf (5)
file.
.TP
.BI \-F \ confdir
specify a config directory.
If both
.B \-f
and
.B \-F
are specified, the config file will be read and converted to
config directory format and written to the specified directory.
If neither option is specified, an attempt to read the
default config directory will be made before trying to use the default
config file. If a valid config directory exists then the
default config file is ignored. If dry-run mode is also specified,
no conversion will occur.
.TP
.B \-g
disable subordinate gluing.  Only the specified database will be
processed, and not its glued subordinates (if any).
.TP
.BI \-j \ lineno
Jump to the specified line number in the LDIF file before processing
any entries. This allows a modification that was aborted due to errors in the
input LDIF to be resumed after the errors are corrected.
.TP
.BI \-l \ ldif-file
Read LDIF from the specified file instead of standard input.
.TP
.BI \-n \ dbnum
Perform changes on the \fIdbnum\fR-th database listed in the
configuration file.  The
.B \-n
cannot be used in conjunction with the
.B \-b
option.
To manipulate the config database
.BR slapd\-config (5),
use
.B \-n 0
as it is always the first database. It must physically exist
on the filesystem prior to this, however.
.TP
.BI \-o \ option\fR[ = value\fR]
Specify an
.I option
with a(n optional)
.IR value .
Possible generic options/values are:
.LP
.nf
              syslog=<subsystems>  (see `\-s' in slapd(8))
              syslog\-level=<level> (see `\-S' in slapd(8))
              syslog\-user=<user>   (see `\-l' in slapd(8))

              schema-check={yes|no}
              value-check={yes|no}

.in
The \fIschema\-check\fR option toggles schema checking (default on);
the \fIvalue\-check\fR option toggles value checking (default off).
The latter is incompatible with \fB-q\fR.
.TP
.B \-q
enable quick (fewer integrity checks) mode.  Does fewer consistency checks
on the input data, and no consistency checks when writing the database.
Improves the run time but if any errors or interruptions occur the resulting
database will be unusable.
.TP
.B \-s
disable schema checking.  This option is intended to be used when
manipulating databases containing special objects, such as fractional
objects on a partial replica.  Creating normal objects which do not
conform to schema may result in unexpected and ill behavior.
.TP
.BI \-S \ SID
Server ID to use in generated entryCSN.  Also used for contextCSN
if \fB\-w\fP is set as well.  Defaults to \fB0\fP.
.TP
.B \-u
enable dry-run (don't write to backend) mode.
.TP
.B \-v
enable verbose mode.
.TP
.BI \-w
write syncrepl context information.
After all entries are added, the contextCSN
will be updated with the greatest CSN in the database.
.SH LIMITATIONS
Your
.BR slapd (8)
should not be running
when you do this to ensure consistency of the database.
.LP
Not all backends support all types of modification, \fImodrdn\fR
changetype in particular is not implemented for any of the current
backends.
.LP
.B slapmodify
may not provide naming or schema checks.  It is advisable to
use
.BR ldapmodify (1)
when possible.
.SH EXAMPLES
To make modifications specified in file
.B ldif
into your
.BR slapd (8)
database give the command:
.LP
.nf
.ft tt
	SBINDIR/slapmodify \-l ldif
.ft
.fi
.SH "SEE ALSO"
.BR ldap (3),
.BR ldif (5),
.BR slapcat (8),
.BR slapadd (8),
.BR slapindex (8),
.BR ldapmodify (1),
.BR slapd (8)
.LP
"OpenLDAP Administrator's Guide" (http://www.OpenLDAP.org/doc/admin/)
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.4 Change Log

OpenLDAP 2.4.50 Engineering
	Contrib
		Update ldapc++ config.guess and config.sub to support newer architectures (ITS#7855)

OpenLDAP 2.4.49 Release (2020/01/30)
	Added slapd-monitor database entry count for slapd-mdb (ITS#9154)
	Fixed client tools to not add controls on cancel/abandon (ITS#9145)
	Fixed client tools SyncInfo message to be LDIF compliant (ITS#8116)
	Fixed libldap to correctly free sb (ITS#9081, ITS#8755)
	Fixed libldap descriptor leak if ldaps fails (ITS#9147)
	Fixed libldap remove unnecessary global mutex for GnuTLS (ITS#9069)
	Fixed slapd syntax evaluation of preferredDeliveryMethod (ITS#9067)
	Fixed slapd to relax domainScope control check (ITS#9100)
	Fixed slapd to have cleaner error handling during connection setup (ITS#9112)
	Fixed slapd data check when processing cancel exop (ITS#9124)
	Fixed slapd attribute description processing (ITS#9128)
	Fixed slapd-ldap to set oldctrls correctly (ITS#9076)
	Fixed slapd-mdb to honor unchecked limit with alias deref (ITS#7657)
	Fixed slapd-mdb missing final commit with slapindex (ITS#9095)
	Fixed slapd-mdb drop attr mappings added in an aborted txn (ITS#9091)
	Fixed slapd-mdb nosync FLAG configuration handling (ITS#9150)
	Fixed slapd-monitor global operation counter reporting (ITS#9119)
	Fixed slapo-ppolicy when used with slapauth (ITS#8629)
	Fixed slapo-ppolicy to add a missed normalised copy of pwdChangedTime (ITS#9126)
	Fixed slapo-syncprov fix sessionlog init (ITS#9146)
	Fixed slapo-unique loop termination (ITS#9077)
	Build Environment
		Fix mkdep to honor TMPDIR if set (ITS#9062)
		Remove ICU library detection (ITS#9144)
		Update config.guess and config.sub to support newer architectures (ITS#7855)
		Disable ITS8521 regression test as it is no longer valid (ITS#9015)
	Documentation
		admin24 - Fix inconsistent whitespace in replication section (ITS#9153)
		slapd-config(5)/slapd.conf(5) - Fix missing bold tag for keyword (ITS#9063)
		slapd-ldap(5) - Document "tls none" option (ITS#9071)
		slapo-ppolicy(5) - Correctly document pwdGraceAuthnLimit (ITS#9065)

OpenLDAP 2.4.48 Release (2019/07/24)
	Added libldap OpenSSL Elliptic Curve support (ITS#7595)
	Added libldap Expose OpenLDAP specific interfaces via openldap.h (ITS#8671)
	Added slapd-monitor support for slapd-mdb (ITS#7770)
	Fixed liblber leaks (ITS#8727)
	Fixed liblber with partial flush (ITS#8864)
	Fixed libldap ASYNC TLS so it works (ITS#8957,ITS#8980)
	Fixed libldap ASYNC connections with Solaris 10 (ITS#8968)
	Fixed libldap with SASL_NOCANON=on and ldapi connections (ITS#7585)
	Fixed libldap to be able to unset syncrepl TLS options (ITS#7042)
	Fixed libldap race condition in ldap_int_initialize (ITS#7996, ITS#8450)
	Fixed libldap return code in ldap_create_assertion_control_value (ITS#8674)
	Fixed libldap to correctly disable IPv6 when configured to do so (ITS#8754)
	Fixed libldap to correctly close TLS connection (ITS#8755)
	Fixed libldap with non-blocking TLS and referals (ITS#8167)
	Fixed libldap_r handling of deprecated OpenSSL function (ITS#8353)
	Fixed liblunicode case correspondance (ITS#8508)
	Fixed slapd with an idletimeout of less than four seconds (ITS#8952)
	Fixed slapd config parser variable for Windows64 (ITS#9012)
	Fixed slapd syncrepl fallback handling with delta-syncrepl (ITS#9015)
	Fixed slapd telephoneNumberNormalize, cert DN validation (ITS#8999)
	Fixed slapd syncrepl for relax with delta-syncrepl (ITS#8037)
	Fixed slapd to restrict rootDN proxyauthz to its own databases (ITS#9038)
	Fixed slapd to initialize SASL SSF per connection (ITS#9052)
	Fixed slapo-accesslog with SLAP_MOD_SOFT modifications (ITS#8990)
	Fixed slapd-ldap starttls connections timeout behavior (ITS#8963)
	Fixed slapd-ldap segfault when entry result doesn't match filter (ITS#8997)
	Fixed slapd-meta conversion from slapd.conf to cn=config (ITS#8743)
	Fixed slapd-meta assertion when network interface goes down (ITS#8841)
	Fixed slapd-mdb fix bitshift integer overflow (ITS#8989)
	Fixed slapd-mdb index cleanup with cn=config (ITS#8472)
	Fixed slapd-mdb to improve performance with alias deref (ITS#7657)
	Fixed slapo-accesslog possible assert with exops (ITS#8971)
	Fixed slapo-chain to correctly reject multiple chaining URIs (ITS#8637)
	Fixed slapo-chain conversion from slapd.conf to cn=config (ITS#8799)
	Fixed slapo-memberof conversion from slapd.conf to cn=config (ITS#8663)
	Fixed slapo-memberof for group name change to itself (ITS#9000)
	Fixed slapo-ppolicy behavior when pwdInHistory is changed (ITS#8349)
	Fixed slapo-rwm to not free original filter (ITS#8964)
	Fixed slapo-syncprov contextCSN generation (ITS#9015)
	Build Environment
		Fixed slapd to only link to BDB libraries with static build (ITS#8948)
		Fixed libldap implicit declaration with LDAP_CONNECTIONLESS (ITS#8794)
		Fixed libldap double inclusion of limits.h in cyrus.c (ITS#9041)
	Documentation
		General - Fixed minor typos (ITS#8764, ITS#8761)
		admin24 - Miscellaneous updates promoting mdb and fixing examples (ITS#9031)
		slapd.access(5) - Note MDB is the primary backend (ITS#8881)
		slapd.backends(5) - Note MDB is the recommended backend (ITS#8771)
		slapd-ldap(5) - Document starttls parameter (ITS#8693)
	Contrib
		Added slapo-lastbind capability to forward authTimestamp updates (ITS#7721)

OpenLDAP 2.4.47 Release (2018/12/19)
	Added slapd-sock DN qualifier for subtrees to be processed (ITS#8051)
	Added slapd-sock ability to send extended operations to external listeners (ITS#8714)
	Fixed liblber to avoid incremental access to user-supplied bv in dupbv (ITS#8752)
	Fixed libldap dn to domain parsing with bad input (ITS#8842)
	Fixed slapd slapcat to correctly honor -g option (ITS#8667)
	Fixed slapd to correctly handle NO_SUCH_OBJECT with dynamic groups (ITS#8923)
	Fixed slapd to check status of rdnNormalize (ITS#8932)
	Fixed slapd cn=config when modifying slapo-syncprov config (ITS#8616)
	Fixed slapd sasl authz-policy "all" behavior (ITS#8909)
	Fixed slapd sasl minor typo (ITS#8918)

servers/slapd/slapmodify.c
/* $OpenLDAP$ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2020 The OpenLDAP Foundation.
 * Portions Copyright 1998-2003 Kurt D. Zeilenga.
 * Portions Copyright 2003 IBM Corporation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* ACKNOWLEDGEMENTS:
 * This work was initially developed by Pierangelo Masarati for inclusion
 * in OpenLDAP Software.
 */

#include "portable.h"

#include <stdio.h>

#include "ac/stdlib.h"

#include "ac/ctype.h"
#include "ac/string.h"
#include "ac/socket.h"
#include "ac/unistd.h"

#include "lber.h"
#include "ldif.h"
#include "lutil.h"
#include "lutil_meter.h"
#include <sys/stat.h>

#include "slapcommon.h"

extern int slap_DN_strict;	/* dn.c */

static char csnbuf[ LDAP_PVT_CSNSTR_BUFSIZE ];

int
slapmodify( int argc, char **argv )
{
	char *buf = NULL;
	const char *text;
	char textbuf[SLAP_TEXT_BUFLEN] = { '\0' };
	size_t textlen = sizeof textbuf;
	const char *progname = "slapmodify";

	struct berval csn;
	unsigned long sid;
	struct berval bvtext;
	ID id;
	OperationBuffer opbuf;
	Operation *op;

	int checkvals, ldifrc;
	unsigned long lineno, nextline;
	int lmax;
	int rc = EXIT_SUCCESS;

	int enable_meter = 0;
	lutil_meter_t meter;
	struct stat stat_buf;

	/* default "000" */
	csnsid = 0;

	if ( isatty (2) ) enable_meter = 1;
	slap_tool_init( progname, SLAPMODIFY, argc, argv );

	memset( &opbuf, 0, sizeof(opbuf) );
	op = &opbuf.ob_op;
	op->o_hdr = &opbuf.ob_hdr;
	op->o_bd = be;

	if ( !be->be_entry_open ||
		!be->be_entry_close ||
		!be->be_entry_put ||
		!be->be_dn2id_get ||
		!be->be_entry_get ||
		!be->be_entry_modify )
	{
		fprintf( stderr, "%s: database doesn't support necessary operations.\n",
			progname );
		if ( dryrun ) {
			fprintf( stderr, "\t(dry) continuing...\n" );

		} else {
			exit( EXIT_FAILURE );
		}
	}

	checkvals = (slapMode & SLAP_TOOL_QUICK) ? 0 : 1;

	lmax = 0;
	nextline = 0;

	/* enforce schema checking unless not disabled and allow unknown
	 * attributes otherwise */
	if ( (slapMode & SLAP_TOOL_NO_SCHEMA_CHECK) == 0) {
		SLAP_DBFLAGS(be) &= ~(SLAP_DBFLAG_NO_SCHEMA_CHECK);
	} else {
		slap_DN_strict = 0;
	}

	if( !dryrun && be->be_entry_open( be, 1 ) != 0 ) {
		fprintf( stderr, "%s: could not open database.\n",
			progname );
		exit( EXIT_FAILURE );
	}

	(void)slap_tool_update_ctxcsn_init();

	if ( enable_meter 
#ifdef LDAP_DEBUG
		/* tools default to "none" */
		&& slap_debug == LDAP_DEBUG_NONE
#endif
		&& !fstat ( fileno ( ldiffp->fp ), &stat_buf )
		&& S_ISREG(stat_buf.st_mode) ) {
		enable_meter = !lutil_meter_open(
			&meter,
			&lutil_meter_text_display,
			&lutil_meter_linear_estimator,
			stat_buf.st_size);
	} else {
		enable_meter = 0;
	}

	/* nextline is the line number of the end of the current entry */
	for( lineno=1; ( ldifrc = ldif_read_record( ldiffp, &nextline, &buf, &lmax )) > 0;
		lineno=nextline+1 )
	{
		BackendDB *bd;
		Entry *e_orig = NULL, *e = NULL;
		struct berval rbuf;
		LDIFRecord lr;
		struct berval ndn = BER_BVNULL;
		int n;
		int is_oc = 0;
		int local_rc;
		int mod_err = 0;
		char *request = "(unknown)";

		ber_str2bv( buf, 0, 0, &rbuf );

		if ( lineno < jumpline )
			continue;

		if ( enable_meter )
			lutil_meter_update( &meter,
					 ftell( ldiffp->fp ),
					 0);

		/*
		 * Initialize text buffer
		 */
		bvtext.bv_len = textlen;
		bvtext.bv_val = textbuf;
		bvtext.bv_val[0] = '\0';

		local_rc = ldap_parse_ldif_record( &rbuf, lineno, &lr,
			"slapmodify", LDIF_NO_CONTROLS );

		if ( local_rc != LDAP_SUCCESS ) {
			fprintf( stderr, "%s: could not parse entry (line=%lu)\n",
				progname, lineno );
			rc = EXIT_FAILURE;
			if( continuemode ) continue;
			break;
		}

		switch ( lr.lr_op ) {
		case LDAP_REQ_ADD:
			request = "add";
			break;

		case LDAP_REQ_MODIFY:
			request = "modify";
			break;

		case LDAP_REQ_DELETE:
			if ( be->be_entry_delete )
			{
				request = "delete";
				break;
			}
			/* backend does not support delete, fallthru */

		case LDAP_REQ_MODRDN:
			fprintf( stderr, "%s: request 0x%lx not supported (line=%lu)\n",
				progname, (unsigned long)lr.lr_op, lineno );
			rc = EXIT_FAILURE;
			goto cleanup;

		default:
			/* record skipped e.g. version: or comment or something we don't handle yet */
			goto cleanup;
		}

		local_rc = dnNormalize( 0, NULL, NULL, &lr.lr_dn, &ndn, NULL );
		if ( local_rc != LDAP_SUCCESS ) {
			fprintf( stderr, "%s: DN=\"%s\" normalization failed (line=%lu)\n",
				progname, lr.lr_dn.bv_val, lineno );
			rc = EXIT_FAILURE;
			goto cleanup;
		}

		/* make sure the DN is not empty */
		if( BER_BVISEMPTY( &ndn ) &&
			!BER_BVISEMPTY( be->be_nsuffix ))
		{
			fprintf( stderr, "%s: line %lu: "
				"%s entry with empty dn=\"\"",
				progname, lineno, request );
			bd = select_backend( &ndn, nosubordinates );
			if ( bd ) {
				BackendDB *bdtmp;
				int dbidx = 0;
				LDAP_STAILQ_FOREACH( bdtmp, &backendDB, be_next ) {
					if ( bdtmp == bd ) break;
					dbidx++;
				}

				assert( bdtmp != NULL );
				
				fprintf( stderr, "; did you mean to use database #%d (%s)?",
					dbidx,
					bd->be_suffix[0].bv_val );

			}
			fprintf( stderr, "\n" );
			rc = EXIT_FAILURE;
			goto cleanup;
		}

		/* check backend */
		bd = select_backend( &ndn, nosubordinates );
		if ( bd != be ) {
			fprintf( stderr, "%s: line %lu: "
				"database #%d (%s) not configured to hold \"%s\"",
				progname, lineno,
				dbnum,
				be->be_suffix[0].bv_val,
				lr.lr_dn.bv_val );
			if ( bd ) {
				BackendDB *bdtmp;
				int dbidx = 0;
				LDAP_STAILQ_FOREACH( bdtmp, &backendDB, be_next ) {
					if ( bdtmp == bd ) break;
					dbidx++;
				}

				assert( bdtmp != NULL );
				
				fprintf( stderr, "; did you mean to use database #%d (%s)?",
					dbidx,
					bd->be_suffix[0].bv_val );

			} else {
				fprintf( stderr, "; no database configured for that naming context" );
			}
			fprintf( stderr, "\n" );
			rc = EXIT_FAILURE;
			goto cleanup;
		}

		/* get id and/or entry */
		switch ( lr.lr_op ) {
			case LDAP_REQ_ADD:
				e = entry_alloc();
				ber_dupbv( &e->e_name, &lr.lr_dn );
				ber_dupbv( &e->e_nname, &ndn );
				break;

			//case LDAP_REQ_MODRDN:
			case LDAP_REQ_DELETE:
			case LDAP_REQ_MODIFY:
				id = be->be_dn2id_get( be, &ndn );
				rc = (id == NOID);
				if ( rc == LDAP_SUCCESS && lr.lr_op != LDAP_REQ_DELETE ) {
					e_orig = be->be_entry_get( be, id );
					if ( e_orig )
						e = entry_dup( e_orig );
					rc = (e == NULL);
				}
				break;
		}

		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "%s: no such entry \"%s\" in database (lineno=%lu)\n",
				progname, ndn.bv_val, lineno );
			rc = EXIT_FAILURE;
			goto cleanup;
		}

		if ( lr.lrop_mods ) {
			for ( n = 0; lr.lrop_mods && lr.lrop_mods[ n ] != NULL; n++ ) {
				LDAPMod *mod = lr.lrop_mods[ n ];
				Modification mods = { 0 };
				unsigned i = 0;
				int bin = (mod->mod_op & LDAP_MOD_BVALUES);
				int pretty = 0;
				int normalize = 0;

				local_rc = slap_str2ad( mod->mod_type, &mods.sm_desc, &text );
				if ( local_rc != LDAP_SUCCESS ) {
					fprintf( stderr, "%s: slap_str2ad(\"%s\") failed for entry \"%s\" (%d: %s, lineno=%lu)\n",
						progname, mod->mod_type, lr.lr_dn.bv_val, local_rc, text, lineno );
					rc = EXIT_FAILURE;
					goto cleanup;
				}

				mods.sm_type = mods.sm_desc->ad_cname;

				if ( mods.sm_desc->ad_type->sat_syntax->ssyn_pretty ) {
					pretty = 1;

				} else {
					assert( mods.sm_desc->ad_type->sat_syntax->ssyn_validate != NULL );
				}

				if ( mods.sm_desc->ad_type->sat_equality &&
					mods.sm_desc->ad_type->sat_equality->smr_normalize )
				{
					normalize = 1;
				}

				if ( bin && mod->mod_bvalues ) {
					for ( i = 0; mod->mod_bvalues[ i ] != NULL; i++ )
						;

				} else if ( !bin && mod->mod_values ) {
					for ( i = 0; mod->mod_values[ i ] != NULL; i++ )
						;
				}

				if ( i != 0 )
				{
					mods.sm_values = SLAP_CALLOC( sizeof( struct berval ), i + 1 );
					if ( normalize ) {
						mods.sm_nvalues = SLAP_CALLOC( sizeof( struct berval ), i + 1 );
					} else {
						mods.sm_nvalues = NULL;
					}
				}
				mods.sm_numvals = i;

				for ( i = 0; i < mods.sm_numvals; i++ ) {
					struct berval bv;

					if ( bin ) {
						bv = *mod->mod_bvalues[ i ];
					} else {
						ber_str2bv( mod->mod_values[ i ], 0, 0, &bv );
					}

					if ( pretty ) {
						local_rc = ordered_value_pretty( mods.sm_desc,
						&bv, &mods.sm_values[i], NULL );

					} else {
						local_rc = ordered_value_validate( mods.sm_desc,
							&bv, 0 );
					}

					if ( local_rc != LDAP_SUCCESS ) {
						fprintf( stderr, "%s: DN=\"%s\": unable to %s attr=%s value #%d\n",
							progname, e->e_dn, pretty ? "prettify" : "validate",
							mods.sm_desc->ad_cname.bv_val, i );
						/* handle error */
						rc = EXIT_FAILURE;
						ber_bvarray_free( mods.sm_values );
						ber_bvarray_free( mods.sm_nvalues );
						goto cleanup;
					}

					if ( !pretty ) {
						ber_dupbv( &mods.sm_values[i], &bv );
					}

					if ( normalize ) {
						local_rc = ordered_value_normalize(
							SLAP_MR_VALUE_OF_ATTRIBUTE_SYNTAX,
							mods.sm_desc,
							mods.sm_desc->ad_type->sat_equality,
							&mods.sm_values[i], &mods.sm_nvalues[i],
							NULL );
						if ( local_rc != LDAP_SUCCESS ) {
							fprintf( stderr, "%s: DN=\"%s\": unable to normalize attr=%s value #%d\n",
								progname, e->e_dn, mods.sm_desc->ad_cname.bv_val, i );
							/* handle error */
							rc = EXIT_FAILURE;
							ber_bvarray_free( mods.sm_values );
							ber_bvarray_free( mods.sm_nvalues );
							goto cleanup;
						}
					}
				}

				mods.sm_op = (mod->mod_op & ~LDAP_MOD_BVALUES);
				mods.sm_flags = 0;

				if ( mods.sm_desc == slap_schema.si_ad_objectClass ) {
					is_oc = 1;
				}

				switch ( mods.sm_op ) {
				case LDAP_MOD_ADD:
					local_rc = modify_add_values( e, &mods,
						0, &text, textbuf, textlen );
					break;

				case LDAP_MOD_DELETE:
					local_rc = modify_delete_values( e, &mods,
						0, &text, textbuf, textlen );
					break;

				case LDAP_MOD_REPLACE:
					local_rc = modify_replace_values( e, &mods,
						0, &text, textbuf, textlen );
					break;

				case LDAP_MOD_INCREMENT:
					local_rc = modify_increment_values( e, &mods,
						0, &text, textbuf, textlen );
					break;
				}

				ber_bvarray_free( mods.sm_values );
				ber_bvarray_free( mods.sm_nvalues );

				if ( local_rc != LDAP_SUCCESS ) {
					fprintf( stderr, "%s: DN=\"%s\": unable to modify attr=%s\n",
						progname, e->e_dn, mods.sm_desc->ad_cname.bv_val );
					rc = EXIT_FAILURE;
					goto cleanup;
				}
			}

			rc = slap_tool_entry_check( progname, op, e, lineno, &text, textbuf, textlen );
			if ( rc != LDAP_SUCCESS ) {
				rc = EXIT_FAILURE;
				goto cleanup;
			}
		}

		if ( SLAP_LASTMOD(be) && e != NULL ) {
			time_t now = slap_get_time();
			char uuidbuf[ LDAP_LUTIL_UUIDSTR_BUFSIZE ];
			struct berval vals[ 2 ];

			struct berval name, timestamp;

			struct berval nvals[ 2 ];
			struct berval nname;
			char timebuf[ LDAP_LUTIL_GENTIME_BUFSIZE ];

			Attribute *a;

			vals[1].bv_len = 0;
			vals[1].bv_val = NULL;

			nvals[1].bv_len = 0;
			nvals[1].bv_val = NULL;

			csn.bv_len = ldap_pvt_csnstr( csnbuf, sizeof( csnbuf ), csnsid, 0 );
			csn.bv_val = csnbuf;

			timestamp.bv_val = timebuf;
			timestamp.bv_len = sizeof(timebuf);

			slap_timestamp( &now, &timestamp );

			if ( BER_BVISEMPTY( &be->be_rootndn ) ) {
				BER_BVSTR( &name, SLAPD_ANONYMOUS );
				nname = name;
			} else {
				name = be->be_rootdn;
				nname = be->be_rootndn;
			}

			a = attr_find( e->e_attrs, slap_schema.si_ad_entryUUID );
			if ( a != NULL ) {
				if ( a->a_vals != a->a_nvals ) {
					SLAP_FREE( a->a_nvals[0].bv_val );
					SLAP_FREE( a->a_nvals );
				}
				SLAP_FREE( a->a_vals[0].bv_val );
				SLAP_FREE( a->a_vals );
				a->a_vals = NULL;
				a->a_nvals = NULL;
				a->a_numvals = 0;
			}
			vals[0].bv_len = lutil_uuidstr( uuidbuf, sizeof( uuidbuf ) );
			vals[0].bv_val = uuidbuf;
			attr_merge_normalize_one( e, slap_schema.si_ad_entryUUID, vals, NULL );

			a = attr_find( e->e_attrs, slap_schema.si_ad_creatorsName );
			if ( a == NULL ) {
				vals[0] = name;
				nvals[0] = nname;
				attr_merge( e, slap_schema.si_ad_creatorsName, vals, nvals );

			} else {
				ber_bvreplace( &a->a_vals[0], &name );
				ber_bvreplace( &a->a_nvals[0], &nname );
			}

			a = attr_find( e->e_attrs, slap_schema.si_ad_createTimestamp );
			if ( a == NULL ) {
				vals[0] = timestamp;
				attr_merge( e, slap_schema.si_ad_createTimestamp, vals, NULL );

			} else {
				ber_bvreplace( &a->a_vals[0], &timestamp );
			}

			a = attr_find( e->e_attrs, slap_schema.si_ad_entryCSN );
			if ( a == NULL ) {
				vals[0] = csn;
				attr_merge( e, slap_schema.si_ad_entryCSN, vals, NULL );

			} else {
				ber_bvreplace( &a->a_vals[0], &csn );
			}

			a = attr_find( e->e_attrs, slap_schema.si_ad_modifiersName );
			if ( a == NULL ) {
				vals[0] = name;
				nvals[0] = nname;
				attr_merge( e, slap_schema.si_ad_modifiersName, vals, nvals );

			} else {
				ber_bvreplace( &a->a_vals[0], &name );
				ber_bvreplace( &a->a_nvals[0], &nname );
			}

			a = attr_find( e->e_attrs, slap_schema.si_ad_modifyTimestamp );
			if ( a == NULL ) {
				vals[0] = timestamp;
				attr_merge( e, slap_schema.si_ad_modifyTimestamp, vals, NULL );

			} else {
				ber_bvreplace( &a->a_vals[0], &timestamp );
			}
		}

		/* check schema, objectClass etc */

		if ( !dryrun ) {
			switch ( lr.lr_op ) {
			case LDAP_REQ_ADD:
				id = be->be_entry_put( be, e, &bvtext );
				rc = (id == NOID);
				break;

			case LDAP_REQ_MODIFY:
				id = be->be_entry_modify( be, e, &bvtext );
				rc = (id == NOID);
				break;

			case LDAP_REQ_DELETE:
				rc = be->be_entry_delete( be, &ndn, &bvtext );
				break;

			}

			if( rc != LDAP_SUCCESS ) {
				fprintf( stderr, "%s: could not %s entry dn=\"%s\" "
					"(line=%lu): %s\n", progname, request, ndn.bv_val,
					lineno, bvtext.bv_val );
				rc = EXIT_FAILURE;
				goto cleanup;
			}

			sid = slap_tool_update_ctxcsn_check( progname, e );

			if ( verbose )
				fprintf( stderr, "%s: \"%s\" (%08lx)\n",
					request, ndn.bv_val, (long) id );
		} else {
			if ( verbose )
				fprintf( stderr, "%s: \"%s\"\n",
					request, ndn.bv_val );
		}

cleanup:;
		ldap_ldif_record_done( &lr );
		SLAP_FREE( ndn.bv_val );
		if ( e ) entry_free( e );
		if ( e_orig ) be_entry_release_w( op, e_orig );
		if ( rc != LDAP_SUCCESS && !continuemode ) break;
	}

	if ( ldifrc < 0 )
		rc = EXIT_FAILURE;

	bvtext.bv_len = textlen;
	bvtext.bv_val = textbuf;
	bvtext.bv_val[0] = '\0';

	if ( enable_meter ) {
		lutil_meter_update( &meter, ftell( ldiffp->fp ), 1);
		lutil_meter_close( &meter );
	}

	if ( rc == EXIT_SUCCESS ) {
		rc = slap_tool_update_ctxcsn( progname, sid, &bvtext );
	}

	ch_free( buf );

	if ( !dryrun ) {
		if ( enable_meter ) {
			fprintf( stderr, "Closing DB..." );
		}
		if( be->be_entry_close( be ) ) {
			rc = EXIT_FAILURE;
		}

		if( be->be_sync ) {
			be->be_sync( be );
		}
		if ( enable_meter ) {
			fprintf( stderr, "\n" );
		}
	}

	if ( slap_tool_destroy())
		rc = EXIT_FAILURE;

	return rc;
}

